### PR TITLE
RaspaBaseWorkChain: enable restarts due to timeout:

### PR DESCRIPTION
**Pull request by yakutovicha**
That includes
* Input plugin: add new exit codes: ERROR_SIMULATION_DID_NOT_START and TIMEOUT

* Output parser: detect when (a) simulation did not start (b) did not reach the end

* Utilities: add add_write_binary_restart and register_error_handler

* BaseRestartWorkChain: exit only if error handler returns ExitCode

* RaspaBaseWorkChain: add _handle_timeout error handler

### CODE REVIEW

**Comment by yakutovicha** [comment]
@danieleongari, here are a few notes on the PR:

* Now RaspaBaseWorkChain always restarts jobs that are not completed due to the timeout.

* Unfortunately, binary restart does NOT restore the whole output, so after the restart, some step information is missing in the output.

* I could not add tests for that, because I have to kill the job somehow. With direct scheduler it is not currently possible, see https://github.com/aiidateam/aiida-core/issues/3717

**Comment by danieleongari** [comment]
> Unfortunately, binary restart does NOT restore the whole output
Well, this will make the parser fail: I think this PR is only acceptable if the work chain is able to merge the previous output to the restarted one
> I could not add tests for that, because I have to kill the job somehow.
I'm confused: I expect the restarter to work for a number of times, if at the end of the calculation the simulation is not finished. Therefore, you should be able to make an example with, e.g., 1000 MC cycles that takes 2 minutes, but set the walltime of the calculation to 1 minute and see it finish prematurely and restarting.

**Comment by yakutovicha** [comment]
> > Unfortunately, binary restart does NOT restore the whole output
> 
> Well, this will make the parser fail: I think this PR is only acceptable if the work chain is able to merge the previous output to the restarted one

Okay, I need to specify what I exactly meant: the part that is not restored is located between "Starting simulation" and "Finishing simulation" which we don't parse anyways.

> 
> > I could not add tests for that, because I have to kill the job somehow.
> 
> I'm confused: I expect the restarter to work for a number of times, if at the end of the calculation the simulation is not finished. Therefore, you should be able to make an example with, e.g., 1000 MC cycles that takes 2 minutes, but set the walltime of the calculation to 1 minute and see it finish prematurely and restarting.

The problem is that (a) to my knowledge raspa does not provide a way to specify the walltime and (b) direct scheduler does not kill the job if the time exceeded (see my issue https://github.com/aiidateam/aiida-core/issues/3717)

**Comment by yakutovicha** [comment]
@danieleongari I managed to add a test for the timeout restart. Please review.

File: examples/workchains/example_base_restart_timeout.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    # Specifying the scheduler options
+    builder.raspa.metadata.options = {
+        "resources": {
+            "num_machines": 1,
+            "num_mpiprocs_per_machine": 1,
+        },
+        "max_wallclock_seconds": 1 * 30 * 60,  # 30 min
+        "withmpi": True,
+        "mpirun_extra_params": ["timeout", "10"],
```
Comment (yakutovicha) [inline examples/workchains/example_base_restart_timeout.py]: Note for myself: add a comment that this is to test walltime

**Comment by danieleongari** [comment]
Working very nice! 
Just a few minor improvements.

File: examples/workchains/example_base_restart_timeout.py (hunk 1 above)
Comment (yakutovicha) [reply examples/workchains/example_base_restart_timeout.py]: already done by @danieleongari

**Comment by danieleongari** [comment]
One comment to leave in this PR: 
let's remember that the header of a restarted calculation will parse the system at restart. 
For example here it started with no Kr molecules but it got one at the restart:
```
$ diff head.data1 head.data2
< Wed Jan 22 18:59:25 2020
---
> Wed Jan 22 18:59:46 2020
29,30c29,30
< 	Degrees of freedom:                        0
< 	Translational Degrees of freedom:          0
---
> 	Degrees of freedom:                        3
> 	Translational Degrees of freedom:          3
211,212c211,212
< Maximum trajectory time:      10.0000000000 [ps]
< Each configuration is used with 5 different initial velocities
---
> Maximum trajectory time:       0.0000000000 [ps]
> Each configuration is used with 0 different initial velocities
1375,1376c1375,1376
< Number of adsorbate molecules    : 0
< Component    0 :    0 molecules
---
> Number of adsorbate molecules    : 1
> Component    0 :    1 molecules
1405c1405
< Pseudo Atoms   28 [      Kr]:    0 atoms
---
> Pseudo Atoms   28 [      Kr]:    1 atoms
1474,1475c1474,1475
< Host/Adsorbate energy:                                        0.00000000
< 	Host/Adsorbate VDW energy:                                    0.00000000
---
> Host/Adsorbate energy:                                     -885.86362622
> 	Host/Adsorbate VDW energy:                                 -885.86362622
1494,1495c1494,1495
< Adsorbate/Adsorbate energy:                                        0.00000000
< 	Adsorbate/Adsorbate VDW energy:                                    0.00000000
---
> Adsorbate/Adsorbate energy:                                       -0.00000000
> 	Adsorbate/Adsorbate VDW energy:                                   -0.00000000
1542,1543c1542,1543
< Total energy:     0.000000000000
< 	Total Van der Waals: 0.000000
---
> Total energy:  -885.863626223338
> 	Total Van der Waals: -885.863626
```
Therefore, there may be some workchains where we want to use `max_iterations=1` to avoid unexpected parsing.

**Pull request merged**